cl/alerts/tasks.py

```diff
@@ -31,57 +32,116 @@ def make_alert_messages(
+                docket_alert = DocketAlert.objects.create(
+                    docket=d, user_id=user_pk
+                )
+            else:
+                first_email, auto_subscribe = True, False
+                docket_alert = DocketAlert.objects.create(
+                    docket=d,
+                    user_id=user_pk,
+                    alert_type=DocketAlert.UNSUBSCRIPTION,
+                )
+            subject_dict["auto_subscribe"] = auto_subscribe
+            subject_dict["first_email"] = first_email
+            subject = subject_template.render(subject_dict).strip()
+            email_context["docket_alert"] = docket_alert
+            email_context["first_email"] = first_email
+            email_context["auto_subscribe"] = auto_subscribe
+            msg = EmailMultiAlternatives(
+                subject=subject,
+                body=txt_template.render(email_context),
+                from_email=settings.DEFAULT_ALERTS_EMAIL,
+                to=[recap_email_user.email],
+                headers={f"X-Entity-Ref-ID": f"docket.alert:{d.pk}"},
+            )
+            html = html_template.render(email_context)
+            msg.attach_alternative(html, "text/html")
+            messages.append(msg)
+
+    if email_addresses:
+        subject = subject_template.render(subject_dict).strip()  # Remove
+        # newlines that editors can insist on adding.
+        for email_address in email_addresses:
+            user = User.objects.get(email=email_address)
+            docket_alert = DocketAlert.objects.get(docket=d, user=user)
+            email_context["docket_alert"] = docket_alert
+            msg = EmailMultiAlternatives(
+                subject=subject,
+                body=txt_template.render(email_context),
+                from_email=settings.DEFAULT_ALERTS_EMAIL,
+                to=[email_address],
+                headers={f"X-Entity-Ref-ID": f"docket.alert:{d.pk}"},
+            )
+            html = html_template.render(email_context)
+            msg.attach_alternative(html, "text/html")
+            messages.append(msg)
 
     return messages
 
 
 # Ignore the result or else we'll use a lot of memory.
 @app.task(ignore_result=True)
-def send_docket_alert(d_pk: int, since: datetime) -> None:
+def send_docket_alert(
+    d_pk: int,
+    since: datetime,
+    first_time_recipients: Optional[list[int]] = None,
+) -> None:
     """Send an alert for a given docket
 
     :param d_pk: The docket PK that was modified
     :param since: If we run alerts, notify users about items *since* this time.
+    :param first_time_recipients: The recipient user_pks if needed to send the
+    first case-user notification email.
     :return: None
     """
+
     email_addresses = (
-        User.objects.filter(docket_alerts__docket_id=d_pk)
+        User.objects.filter(
+            docket_alerts__docket_id=d_pk,
+            docket_alerts__alert_type=DocketAlert.SUBSCRIPTION,
+        )
```

Comment:
If you update this filter to be the combination of the query it has now *plus* the first time users, then you can vastly simplify your make_alert_messages code into a single loop.

So the query would be something like, all users:

 - That have a docket alert of type SUBSCRIPTION for this docket, OR
 - Users that have an email address in the `first_time_recipients` list.

Then, in make_alert_messages, you have a single loop:

```
for email_address in email_addresses:
    user = User.objects.get(email=email_address)
    if user.email in first_time_recipients:
        # make the alert, adjust context variables
    # make the message
```

I think that'll make this a LOT simpler.

Reply:
Thanks, I did these changes and allowed us to delete duplicate code and it's more clear.